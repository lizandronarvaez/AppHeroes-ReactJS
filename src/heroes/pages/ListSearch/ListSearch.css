.heroe-is-found{
    width: 30rem;
    margin: 0 auto;
    margin-top: 7rem;
}

.heroe-not-found{
    text-align: center;
    margin-top: 5rem;
}

.heroe-not-found h3 span{
    font-style: italic;
    color: rgb(233, 214, 214);
}

.heroe-not-found button {
    padding: 1rem 2.3rem;
    border:none;
    background-color: rgb(233, 68, 68);
    color:rgb(233, 214, 214) ;
    font-weight: 600;
    margin-top: 3rem;
    border-radius: .5rem;
    font-size: 1.6rem;
}

.heroe-not-found button:hover{
    background-color: rgb(224, 40, 40)
}