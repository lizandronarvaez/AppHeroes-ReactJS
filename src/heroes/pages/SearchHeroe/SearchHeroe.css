.form-busqueda{
    width: 100%;
    text-align: center;
    padding: 0 2rem;
}
.form-busqueda input{
    outline: none;
    margin-right: 3rem;
    padding: .6rem 1rem;
    border-radius: .5rem;
    margin: 0 auto;
    width: 100%;
}

@media (min-width:768px){
    .form-busqueda{
        width: auto;
        padding: 0;
    }
    .form-busqueda input{
        margin-right: 3rem;
        padding: .2rem 1rem;
        margin: 0 auto;
    }
    
}