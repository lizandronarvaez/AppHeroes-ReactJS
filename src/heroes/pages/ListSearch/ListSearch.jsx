// import HeroeCard from "../components/HeroeCard/HeroeCard"
import { useLocation, useNavigate } from "react-router-dom"
import queryString from "query-string";
import HeroeCard from "../../components/HeroeCard/HeroeCard";
import getHeroesByName from "../../helpers/getHeroesByName";
import "./ListSearch.css"
import { useEffect } from "react";
const ListSearch = () => {
    const navigate = useNavigate()
    const location = useLocation();
    // Query del valor de la busqueda
    const { q = "" } = queryString.parse(location.search);
    // Resultado de la busqueda del query
    const heroe = getHeroesByName(q);
    // Volver atras
    const handleBackSearch = () => navigate(-1)

    useEffect(() => {
        q === "" ? navigate("/") : null
    }, [navigate, q])
    return (

        heroe ? (
            <div className="heroe-is-found">
                <HeroeCard
                    key={heroe.id}
                    {...heroe}
                />
            </div>

        ) : (
            <div className="heroe-not-found">
                <h3>No existe el personaje con el nombre <span>{q}</span></h3>
                <button
                    onClick={handleBackSearch}
                >
                    Volver
                </button>
            </div>
        )
    )
}

export default ListSearch