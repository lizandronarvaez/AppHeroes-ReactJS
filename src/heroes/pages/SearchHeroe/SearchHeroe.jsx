import "./SearchHeroe.css"
import useForm from "../../hooks/useForm"
import { useLocation, useNavigate } from "react-router-dom"
import queryString from "query-string";
const SearchHeroe = () => {
    const location = useLocation();
    const navigate = useNavigate();
    const { q = "" } = queryString.parse(location.search);

    const { heroe, handleInputChange, setFormData } = useForm({ heroe: q })

    const handleSubmitSearch = (e) => {
        e.preventDefault();
        const dataInputClean = heroe.toLowerCase().trim();
        if (dataInputClean.length <= 1) return;

        navigate(`heroe/search?q=${dataInputClean}`)
        setFormData({ heroe: "" })
    }
    return (
        <form className="form-busqueda" onSubmit={handleSubmitSearch} aria-label="buscar-personaje">
            <input
                aria-label="valor-personaje"
                type="text"
                name="heroe"
                placeholder="Buscar un personaje"
                onChange={handleInputChange}
                autoComplete="off"
                value={heroe}
            />
        </form>
    )
}

export default SearchHeroe