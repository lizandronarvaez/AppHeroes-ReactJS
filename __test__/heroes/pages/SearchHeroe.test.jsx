/** @jest-environment jsdom */
/* eslint-disable no-undef */
import { fireEvent, render, screen } from "@testing-library/react";
import React from "react"
import SearchHeroe from "../../../src/heroes/pages/SearchHeroe/SearchHeroe";
import { MemoryRouter } from "react-router-dom";
import ListSearch from "../../../src/heroes/pages/ListSearch/ListSearch";
global.React = React;

// ESTA ES UNA FUNCION MOCK PARA LLAMAR A USENAVIGATE
// de esta forma podremos utilizar navigate en las pruebas
const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom')),
    useNavigate: () => mockUseNavigate,
}))
// 

describe('Pruebas en la pagina SearchHeroe', () => {
    beforeEach(() => jest.clearAllMocks);

    test('Debe de mostrar los valores por defecto', () => {

        const { container } = render(
            <MemoryRouter>
                <SearchHeroe />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();
        // screen.debug()
    });


    test('Debe de mostrar a un personaje y su valor', () => {

        const inputValue = "hulk"
        const { container } = render(
            <MemoryRouter initialEntries={[`/search?q=${inputValue}`]} >
                <SearchHeroe />
                <ListSearch />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();

        // comprueba que el valor del input es igual
        const input = screen.getByRole("textbox");
        expect(input.value).toBe(inputValue);
        // Comprueba que la imagen es igual a la imagen que buscamos
        const image = screen.getByRole("img");
        expect(image.src).toContain(`heroes/marvel-${inputValue}`);

        // screen.debug()
    });

    test('Si el personaje existe, debe devolver un div con una clase isFound', () => {

        const inputValue = "hulk"
        const { container } = render(
            <MemoryRouter initialEntries={[`/search?q=${inputValue}`]} >
                <SearchHeroe />
                <ListSearch />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();

        expect(container.getElementsByClassName("heroe-is-found")).toBeTruthy();
        // screen.debug()
    });

    test('Debe de mostrar un error si no encuentro el heroe', () => {

        const inputValue = "loki"
        const { container } = render(
            <MemoryRouter initialEntries={[`/search?q=${inputValue}`]} >
                <SearchHeroe />
                <ListSearch />
            </MemoryRouter>
        )
        expect(container).toMatchSnapshot();

        expect(container.getElementsByClassName("heroe-not-found").length).toBe(1);

        // screen.debug()
    });

    test('Realizar la busqueda en el formulario nos lleva a otra pagina con el useNavigate()', () => {

        const inputValue = "loki"

        const { container } = render(
            <MemoryRouter initialEntries={["/"]}>
                <SearchHeroe />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot();

        const inputBusqueda = screen.getByLabelText("valor-personaje")
        fireEvent.change(inputBusqueda, { target: { name: "heroe", value: inputValue } })

        const formulario = screen.getByLabelText("buscar-personaje");
        fireEvent.submit(formulario)

        expect(mockUseNavigate).toHaveBeenCalledWith(`heroe/search?q=${inputValue}`);
        // screen.debug()
    });
});