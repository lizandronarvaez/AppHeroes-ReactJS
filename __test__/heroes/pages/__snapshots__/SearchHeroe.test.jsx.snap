// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en la pagina SearchHeroe Debe de mostrar a un personaje y su valor 1`] = `
<div>
  <form
    aria-label="buscar-personaje"
    class="form-busqueda"
  >
    <input
      aria-label="valor-personaje"
      autocomplete="off"
      name="heroe"
      placeholder="Buscar un personaje"
      type="text"
      value="hulk"
    />
  </form>
  <div
    class="heroe-is-found"
  >
    <div
      class="card"
    >
      <img
        alt="Hulk"
        class="card-img"
        src="/heroes/marvel-hulk.jpg"
      />
      <div
        class="card-body text-center"
      >
        <h5>
          Bruce Banner
        </h5>
        <p>
          Hulk
        </p>
        <a
          href="/heroe/marvel-hulk"
        >
          Informacion
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en la pagina SearchHeroe Debe de mostrar los valores por defecto 1`] = `
<div>
  <form
    aria-label="buscar-personaje"
    class="form-busqueda"
  >
    <input
      aria-label="valor-personaje"
      autocomplete="off"
      name="heroe"
      placeholder="Buscar un personaje"
      type="text"
      value=""
    />
  </form>
</div>
`;

exports[`Pruebas en la pagina SearchHeroe Debe de mostrar un error si no encuentro el heroe 1`] = `
<div>
  <form
    aria-label="buscar-personaje"
    class="form-busqueda"
  >
    <input
      aria-label="valor-personaje"
      autocomplete="off"
      name="heroe"
      placeholder="Buscar un personaje"
      type="text"
      value="loki"
    />
  </form>
  <div
    class="heroe-not-found"
  >
    <h3>
      No existe el personaje con el nombre 
      <span>
        loki
      </span>
    </h3>
    <button>
      Volver
    </button>
  </div>
</div>
`;

exports[`Pruebas en la pagina SearchHeroe Realizar la busqueda en el formulario nos lleva a otra pagina con el useNavigate() 1`] = `
<div>
  <form
    aria-label="buscar-personaje"
    class="form-busqueda"
  >
    <input
      aria-label="valor-personaje"
      autocomplete="off"
      name="heroe"
      placeholder="Buscar un personaje"
      type="text"
      value=""
    />
  </form>
</div>
`;

exports[`Pruebas en la pagina SearchHeroe Si el personaje existe, debe devolver un div con una clase isFound 1`] = `
<div>
  <form
    aria-label="buscar-personaje"
    class="form-busqueda"
  >
    <input
      aria-label="valor-personaje"
      autocomplete="off"
      name="heroe"
      placeholder="Buscar un personaje"
      type="text"
      value="hulk"
    />
  </form>
  <div
    class="heroe-is-found"
  >
    <div
      class="card"
    >
      <img
        alt="Hulk"
        class="card-img"
        src="/heroes/marvel-hulk.jpg"
      />
      <div
        class="card-body text-center"
      >
        <h5>
          Bruce Banner
        </h5>
        <p>
          Hulk
        </p>
        <a
          href="/heroe/marvel-hulk"
        >
          Informacion
        </a>
      </div>
    </div>
  </div>
</div>
`;
